sourcetable.go

```diff
@@ -123,3 +127,163 @@ func (m MountEntry) String() string {
 		m.CountryCode, m.Latitude, m.Longitude, nmea, solution, m.Generator, m.Compression,
 		m.Authentication, fee, m.Bitrate, m.Misc)
 }
+
+// DecodeSourcetable parses a sourcetable from an ioreader into a ntrip style source table.
+func DecodeSourcetable(in []byte) (Sourcetable, []error) {
+	table := Sourcetable{}
+	str := string(in[:])
+	var errs []error
+
+	lines := strings.Split(str, "\n")
+
+	for lineNo, rawLine := range lines {
+		line := strings.TrimSpace(rawLine)
+
+		if line == "" {
+			continue
+		}
+
+		if line == "ENDSOURCETABLE" {
+			break
+		}
+
+		switch line[:3] {
+		case "CAS":
+			caster, err := buildCasterEntry(line)
+			if err != nil {
+				errs = append(errs, errors.Wrapf(err, "parsing caster %v", lineNo))
+			}
+			table.Casters = append(table.Casters, caster)
+		case "NET":
+			net, err := buildNetEntry(line)
+			if err != nil {
+				errs = append(errs, errors.Wrapf(err, "parsing line %v", lineNo))
+			}
+			table.Networks = append(table.Networks, net)
+		case "STR":
+			mount, err := buildMountEntry(line)
+			if err != nil {
+				errs = append(errs, errors.Wrapf(err, "parsing line %v", lineNo))
+			}
+			table.Mounts = append(table.Mounts, mount)
+		}
+
+	}
+
+	return table, errs
+}
+
+func buildCasterEntry(casterString string) (CasterEntry, error) {
+	parts := strings.Split(casterString, ";")
+
+	long, err := strconv.ParseFloat(parts[8], 64)
+	if err != nil {
+		return CasterEntry{}, fmt.Errorf("invalid longitude")
+	}
+
+	lat, err := strconv.ParseFloat(parts[7], 64)
+	if err != nil {
+		fmt.Println(err)
+		return CasterEntry{}, fmt.Errorf("invalid latitude")
+	}
+
+	nmea := true
+	if parts[5] == "0" {
+		nmea = false
+	}
+
+	port, err := strconv.ParseInt(parts[2], 10, 64)
+	if err != nil {
+		fmt.Println(err)
+		return CasterEntry{}, fmt.Errorf("invalid port")
+	}
+
+	return CasterEntry{
+		Host:       parts[1],
+		Port:       int(port),
+		Identifier: parts[3],
+		Operator:   parts[4],
+		NMEA:       nmea,
+		Country:    parts[6],
+		Latitude:   float32(lat),
+		Longitude:  float32(long),
+	}, nil
+
+}
+
+func buildNetEntry(netString string) (NetworkEntry, error) {
```

Comment:
The naming for these functions is a little inconsistent. Above uses Caster, this is Net (network abbreviated), and below uses Mount.

As I write this I realize that the Sourcetable spec says they are called "Stream records" not Mount records. Where I stream record gives you the mountpoint (the URL path, I suppose).

In light of that I think we should change the name of MountEntry to StreamEntry, and call these methods: ParseCasterEntry, ParseNetworkEntry, and ParseStreamEntry.

Could use Record instead of Entry, but both make sense to me.

Reply:
https://software.rtcm-ntrip.org/wiki/Sourcetable

Reply:
Alright I'll do that now.

Reply:
Is changing MountEntry to StreamEntry going to break stuff elsewhere? We are changing the module API.

Reply:
Nothing of consequence outside of this repository. Everything which uses this will uses modules and so won't change until it's deliberate. We can bump the release now or later.